geo/src/algorithm/dimensions.rs

```diff
@@ -0,0 +1,452 @@
+use crate::{
+    CoordinateType, Geometry, GeometryCollection, Line, LineString, MultiLineString, MultiPoint,
+    MultiPolygon, Point, Polygon, Rect, Triangle,
+};
+
+/// Geometries can have 0, 1, or two dimensions. Or, in the case of an [`empty`](#is_empty)
+/// geometry, a special `Empty` dimensionality.
+///
+/// ## Examples
+///
+/// ```
+/// use geo_types::{Point, Rect, line_string};
+/// use geo::algorithm::dimensions::{HasDimensions, Dimensions};
+///
+/// let point = Point::new(0.0, 5.0);
+/// let line_string = line_string![(x: 0.0, y: 0.0), (x: 5.0, y: 5.0), (x: 0.0, y: 5.0)];
+/// let rect = Rect::new((0.0, 0.0), (10.0, 10.0));
+/// assert_eq!(Dimensions::ZeroDimensional, point.get_dimensions());
+/// assert_eq!(Dimensions::OneDimensional, line_string.get_dimensions());
+/// assert_eq!(Dimensions::TwoDimensional, rect.get_dimensions());
+///
+/// assert!(point.get_dimensions() < line_string.get_dimensions());
+/// assert!(rect.get_dimensions() > line_string.get_dimensions());
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd)]
+pub enum Dimensions {
+    /// Some geometries, like a `MultiPoint` or `GeometryColletion` may have no elements - thus no
+    /// dimensions. Note that this is distinct from being `ZeroDimensional`, like a `Point`.
+    Empty,
```

Comment:
I considered ommitting this and instead representing Empty as a `None` `Option<Dimensions>`, but it seemed to work better with `Empty` as an explicit member of the enum. 

Primarily because I want to use it in comparisons:

```
if a.get_dimensions() < Dimensions::OneDimensional {
    do_something()
} else {
    do_something_else()
}
```

is preferable to something like:

```
match a.get_dimensions() {
    Some(dimension) => {
        if dimensions < Dimensions::OneDimensional {
            do_something()
        } else {
            do_something_else()
        } 
    }
    None => {  do_something_else() }
}
```